src/main/java/com/minecolonies/coremod/entity/ai/util/RecipeStorage.java

```diff
@@ -0,0 +1,123 @@
+package com.minecolonies.coremod.entity.ai.util;
+
+import com.minecolonies.coremod.util.InventoryUtils;
+import net.minecraft.item.ItemStack;
+import net.minecraftforge.items.IItemHandler;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+/**
+ * Class used to represent a recipe in minecolonies.
+ */
+public class RecipeStorage
+{
+    /**
+     * Input required for the recipe.
+     */
+    private final List<ItemStack> input;
+
+    /**
+     * Primary output generated for the recipe.
+     */
+    private final ItemStack primaryOutput;
+
+    /**
+     * Secondary output generated for the recipe.
+     */
+    private final List<ItemStack> secondaryOutput;
+
+    /**
+     * Grid size required for the recipe.
+     */
+    private final int gridSize;
+
+    /**
+     * Create an instance of the recipe storage.
+     * @param input the list of input items (required for the recipe).
+     * @param gridSize the required grid size to make it.
+     * @param primaryOutput the primary output of the recipe.
+     * @param secondaryOutput the secondary output (like buckets or similar).
+     */
+    public RecipeStorage(final List<ItemStack> input, final int gridSize, final ItemStack primaryOutput, final ItemStack...secondaryOutput)
+    {
+        this.input = new ArrayList<>(input);
+        this.primaryOutput = primaryOutput;
+        this.secondaryOutput = new ArrayList<>(Arrays.asList(secondaryOutput));
+        this.gridSize = gridSize;
+    }
+
+    /**
+     * Get the list of input items.
+     * Make a real copy of the itemStack so we're able to request and handle them without changing the recipe.
+     * @return the copy of the list.
+     */
+    public List<ItemStack> getInput()
```

Comment:
The problem seems to be that the checkOrRequest modifies the copies I send them.
So I thought it is probably better to make the copy of the list at one place and not everytime I might request a part of the recipe.